cpu/stm32/i2c: fix wrong speed parameters

aabadie opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md.
-->

### Contribution description

This PR is a take over from #9804 and adapted to new STM32 families added and to the different refactoring done in between.

The main idea is to not disable HSI during clock initialization when I2C is used. I think this saves a few ROM and is a bit faster during initialization compared to the initial approach where this is done in the periph initialization (so enable hsi -> disable hsi -> reenable hsi, whereas with this PR: enable hsi once).

Maybe other families that depends on this version of the I2C driver (f0, f3, l0 and l1) could be adapted but I didn't check yet.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

- Green CI
- I2C is working (probably better)

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

Take over from #9804, so closes #9804

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

fjmolinas commented:
@MrKevinWeiss can the HIL be ran here?

MrKevinWeiss commented:
It can, but I don't have clock checking for the i2c tests yet so I don't know how useful it would be.

MrKevinWeiss commented:
Maybe something for @MarcelStenzel to add?

benpicco commented:
Murdock found some issues

benpicco commented:
@MrKevinWeiss I think it would suffice to verify that I2C still works at all with this change :wink:

MrKevinWeiss commented:
Maybe [this](https://hil.riot-os.org/results/robot_tests/187/overview) is better... Compared to [nightlies](https://hil.riot-os.org/results/nightly/latest/overview) I don't see much difference at least for the nucleos.

fjmolinas commented:
> Maybe [this](https://hil.riot-os.org/results/robot_tests/187/overview) is better... Compared to [nightlies](https://hil.riot-os.org/results/nightly/latest/overview) I don't see much difference at least for the nucleos.

Does that mean master is broken?

fjmolinas commented:
I don't have many `stm32` boards at hand, any chance you could give a few a test @aabadie?

aabadie commented:
Before someone ACKs this PR, I'd like to mention that the proposed approach has a power consumption penalty: on STM32 boards that provides an HSE, when using I2C, the HSI will always be running when not sleeping while in master it's shutdown.

IMHO, the ideal solution would be to compute these parameters once (there's a formula in the manual, but well) based on the bus clock frequency. That will increase RAM/ROM usage but it will guarantee that HSI is not running when not used as sysclk source (either directly or via PLL).

aabadie commented:
Tested on G4, L4 and F7. They are working (I had to push a fix for F7 though).

aabadie commented:
Now it would be good to test on G0, L5, WB and WL, to be complete.

aabadie commented:
Tested on G0, L5, WB and WL: reading I2C sensor values works like a charm.

fjmolinas reviewed (approved):
ACK, trusting @aabadie testing.

benpicco commented:
@aabadie I think you are the best to make the trade-off wrt the HSI question here. So previously HSI was only running when the bus was acquired, now it's always running?
How much does this affect the power draw? And why can't we only enable HSI on demand anymore with this?

aabadie commented:
> So previously HSI was only running when the bus was acquired, now it's always running?

No, before the I2C was simply clocked by SYSCLK which is by default not corresponding to the configured I2C parameters.

> How much does this affect the power draw? And why can't we only enable HSI on demand anymore with this?

I can try to monitor power consumption but I'm not an expert here.
We could enable HSI on demand, but I fear this would a little performance penalty when acquiring/releasing the bus.

aabadie commented:
Hmm, and this needs a rebase

aabadie commented:
Rebased this PR and moved the HSI enable/disable strategy to the acquire/release function of the I2C peripheral driver. I made quick tests on L4 and L0 (HSI always enabled on the L0 board I tried) and the behavior is correct: HSI is enabled if not already enabled and only disabled if the HSI was off before acquiring the bus. Since an I2C bus communication is synchronous, using a static variable for the state should be safe but I need someone else's opinion on this point.

I think some changes are still missing in this PR: I2C configs must be adapted on more boards but I think Murdock will catch them.

fjmolinas commented:
@MrKevinWeiss this seems to be the case for some HIL in the loop testing (only for stm32 BOARDS), could you run some tests?

MrKevinWeiss commented:
> @MrKevinWeiss this seems to be the case for some HIL in the loop testing (only for stm32 BOARDS), could you run some tests?

Sure thing, [here is the current run](https://hil.riot-os.org/jenkins/blue/organizations/jenkins/robot_tests/detail/robot_tests/189/pipeline).  I am am only running the i2c tests from robot fw, the PR feedback will view it as a single test (even though there are many) and if there is a failure it will say the whole thing failed...  I will post a link to easier to read test results once complete.

MrKevinWeiss commented:
[Here are the results](https://hil.riot-os.org/results/robot_tests/189/overview).

So far it looks similar to master, the only difference looks like the `nucleo-l432kc` is locked up, I will give it a retry since it works on master.

MrKevinWeiss commented:
The rerun of the `nucleo-l432kc` works, it seems it was an infra problem.  No difference between master and this PR for the robot tests.

We don't yet have a clock speed metrics for the i2c, maybe @MarcelStenzel would like to add it to his list :wink:

fjmolinas commented:
> Rebased this PR and moved the HSI enable/disable strategy to the acquire/release function of the I2C peripheral driver. I made quick tests on L4 and L0 (HSI always enabled on the L0 board I tried) and the behavior is correct: HSI is enabled if not already enabled and only disabled if the HSI was off before acquiring the bus. Since an I2C bus communication is synchronous, using a static variable for the state should be safe but I need someone else's opinion on this point.
> 
> I think some changes are still missing in this PR: I2C configs must be adapted on more boards but I think Murdock will catch them.

I think thats fine:

- stm32 is only 32 bit, so every read/write should be atomic
- they are done after the mutex, so there can't me multiple accesorss
- AFAIU the mutex call results in a memory barrier

fjmolinas commented:
@aabadie if murdock is going to catch all missing adaptions please squash once thats done

fjmolinas reviewed (approved):
ACK, HIL testing provided, @aabadie did some testing as well.

aabadie commented on cpu/stm32/include/cpu_conf_stm32_common.h:
```diff
@@ -66,6 +62,31 @@ static const i2c_timing_param_t timing_params[] = {
         .sdadel = 0x0,      /* t_SDADEL = 0ns   */
         .scldel = 0x1,      /* t_SCLDEL = 250ns */
     }
+#elif defined(CPU_FAM_STM32L4) || defined(CPU_FAM_STM32F7) || \
```
Open question: should I also include f0 and f3 in this list ? And then drop f0 and f3 specific parameters. I haven't tried I2C on these platforms with this PR and with these parameters but I think that's worth.

fjmolinas replied:
Please do

aabadie replied:
Done, with some extra cleanups: the `cpu_conf_common.h` was not needed anymore. I moved its content in `cpu_i2c.h`, guarded by the same CPU family check and removed the file.
Let's see if that passes Murdock.

fjmolinas commented:
ACK still stands

aabadie commented:
> please squash

Done
